modules/miner/splitsetheap.go

```diff
@@ -0,0 +1,211 @@
+package miner
+
+// mapElements are stored in a mapHeap. The index refers to the location of the
+// splitSet in the underlying slice used to represent the heap.
+type mapElement struct {
+	set   *splitSet
+	id    splitSetID
+	index int
+}
+
+// MapHeap is a heap of splitSets (compared by averageFee). The minHeap bool
+// specifies whether it is a min-heap or max-heap.
+type mapHeap struct {
+	selectID map[splitSetID]*mapElement
+	data     []*mapElement
+	size     uint64
+	minHeap  bool
+}
+
+// up maintains the heap condition by checking if the element at index j is less
+// than its parent (as defined by Less()). If so it swaps them, so that the
+// element at index j goes 'up' the heap. It continues until the heap condition
+// is satisfied again.
+func (mh *mapHeap) up(j int) {
+	for {
+		// i is the parent of element at index j.
+		i := (j - 1) / 2
+
+		if i == j || !mh.Less(j, i) {
+			// Heap condition maintained.
+			break
+		}
+
+		// Swap i and j, then continue.
+		mh.Swap(i, j)
+		j = i
+	}
+}
+
+// down maintains the heap condition by checking that the children of the
+// element at index i are less than the element at i (as defined by Less()). If
+// so, it swaps them, and continues down the heap until the heap condition is
+// satisfied.
+func (mh *mapHeap) down(i0, n int) bool {
+	i := i0
+	for {
+		// j1 is the left child of the element at index i
+		j1 := 2*i + 1
+
+		// Check that j1 is in the bounds of the heap (j1 < 0 after int overflow).
+		if j1 >= n || j1 < 0 {
+			break
+		}
+
+		//j is the left child of i.
+		j := j1
+
+		// If the right child (j2) of the element at index i (the sibling of j),
+		// is within the bounds of the heap and satisfies
+		if j2 := j1 + 1; j2 < n && !mh.Less(j1, j2) {
+
```

Comment:
I realize that you just copied this code, but this newline is weird and should be removed.